RaftMongoWithRaftReconfig.tla

```diff
@@ -0,0 +1,256 @@
+       \/ /\ Len(log[i]) <= Len(log[j])
+          /\ LastTerm(log[i]) /= LogTerm(j, Len(log[i]))
+
+RollbackOplog(i, j) ==
+    /\ CanRollbackOplog(i, j)
+    \* Rollback 1 oplog entry
+    /\ LET new == [index2 \in 1..(Len(log[i]) - 1) |-> log[i][index2]]
+         IN log' = [log EXCEPT ![i] = new]
+    /\ UNCHANGED <<serverVars, committedEntries, configs>>
+
+\* The set of nodes that has log[me][logIndex] in their oplog
+Agree(me, logIndex) ==
+    { node \in ServerViewOn(me) :
+        /\ Len(log[node]) >= logIndex
+        /\ LogTerm(me, logIndex) = LogTerm(node, logIndex) }
+
+NotBehind(me, j) == \/ LastTerm(log[me]) > LastTerm(log[j])
+                    \/ /\ LastTerm(log[me]) = LastTerm(log[j])
+                       /\ Len(log[me]) >= Len(log[j])
+
+\* ACTION
+\* i = the new primary node.
+BecomePrimaryByMagic(i, ayeVoters) ==
+    /\ \A j \in ayeVoters : /\ i \in ServerViewOn(j)
+                            /\ NotBehind(i, j)
+                            /\ currentTerm[j] <= currentTerm[i]
+    /\ ayeVoters \in Quorum(i)
+    /\ state' = [index \in Server |-> IF index \notin ayeVoters
+                                      THEN state[index]
+                                      ELSE IF index = i THEN Leader ELSE Follower]
+    /\ currentTerm' = [index \in Server |-> IF \/ index \in ayeVoters 
+                                               \/ index = i
+                                            THEN currentTerm[i] + 1 
+                                            ELSE currentTerm[index]]
+    /\ LET noop == [term  |-> currentTerm[i] + 1, configVersion |-> GetConfigVersion(i)]
+           newLog == Append(log[i], noop)
+       IN  log' = [log EXCEPT ![i] = newLog]
+    /\ UNCHANGED <<committedEntries, configs>>
+
+\* ACTION
+\* Leader i receives a client request to add v to the log.
+ClientWrite(i) ==
+    /\ state[i] = Leader
+    /\ LET entry == [term  |-> currentTerm[i], configVersion |-> GetConfigVersion(i)]
+           newLog == Append(log[i], entry)
+       IN  log' = [log EXCEPT ![i] = newLog]
+    /\ UNCHANGED <<serverVars, committedEntries, configs>>
+        
+\* ACTION
+AdvanceCommitPoint ==
+    \E leader \in Server : \E acknowledgers \in SUBSET Server :
+        /\ state[leader] = Leader
+        /\ acknowledgers \subseteq Agree(leader, Len(log[leader]))
+        /\ acknowledgers \in Quorum(leader)
+        \* If we comment out the following line, a replicated log entry from old primary will voilate the safety.
+        \* [ P (2), S (), S ()]
+        \* [ S (2), S (), P (3)]
+        \* [ S (2), S (2), P (3)] !!! the log from term 2 shouldn't be considered as committed.
+        /\ LogTerm(leader, Len(log[leader])) = currentTerm[leader]
+        \* If an acknowledger has a higher term, the leader would step down.
+        /\ \A j \in acknowledgers : currentTerm[j] <= currentTerm[leader]
+        /\ committedEntries' = committedEntries \union {[term |-> LastTerm(log[leader]), index |-> Len(log[leader])]}
```

Comment:
The way we track committed entries here may be slightly inaccurate. We are marking the latest log entry of this primary as committed and adding that entry to the `committedEntries` set, but we are not including any entries that may have been prefix committed.

Reply:
I think this model is sufficient to check the invariant that writes that are acknowledged as majority committed never roll back. If we try to include prefix committed entries in the set, I think we will run into https://jira.mongodb.org/browse/SERVER-39626.

Reply:
Good idea, I added a comment above the variable declaration.